torch/csrc/api/src/optim/adagrad.cpp

```diff
@@ -11,42 +12,142 @@
+      }
+      const auto clr = options.lr() /
+          (1 + (state.step() - 1) * options.lr_decay());
+
+      if (grad.is_sparse()) {
+        grad = grad.coalesce();
+        auto grad_indices = grad._indices();
+        auto grad_values = grad._values();
+        auto size = grad.sizes();
+
+        auto make_sparse = [&] (const Tensor& values) -> Tensor {
+          if (grad_indices.dim() == 0 || values.dim() == 0) {
+            return torch::empty({0}, grad.options()).resize_as_(grad);
+          }
+          return torch::sparse_coo_tensor(grad_indices, values, size, grad.options());
+        };
+        state.sum(state.sum().add_(make_sparse(grad_values.pow(2))));
+        auto std = state.sum().sparse_mask(grad);
+        const auto std_values = std._values().sqrt_().add_(options.eps());
+
+        p.data().add_(make_sparse(grad_values / std_values), -clr);
+      }
+      else {
+        state.sum(state.sum().addcmul_(grad, grad, 1.0));
+        const auto std = state.sum().sqrt().add_(options.eps());
+        p.data().addcdiv_(grad, std, -clr);
+      }
     }
+  }
+}
 
-    buffer_at(step_buffers, i) += 1.0;
-    const auto clr = options.learning_rate() /
-        (1.0 + (buffer_at(step_buffers, i) - 1.0) * options.lr_decay());
+void Adagrad::add_parameters(const std::vector<Tensor>& parameters) {
+  param_groups_.emplace_back(OptimizerParamGroup(parameters, defaults_->clone()));
+}
 
-    auto& sum = buffer_at(sum_buffers, i);
-    sum.addcmul_(p.grad(), p.grad(), 1.0);
-    const auto std = buffer_at(sum_buffers, i).sqrt().add_(1e-10);
+const std::vector<Tensor>& Adagrad::parameters() const noexcept {
+  return param_groups_[0].params();
+}
 
-    NoGradGuard guard;
-    p.addcdiv_(p.grad(), std, -clr);
+std::vector<Tensor>& Adagrad::parameters() noexcept {
+  return param_groups_[0].params();
+}
+
+size_t Adagrad::size() const noexcept {
+  size_t count = 0;
+  for (const auto& group : param_groups_) {
+    count += group.params().size();
   }
+  return count;
 }
 
 void Adagrad::save(serialize::OutputArchive& archive) const {
   serialize(*this, archive);
 }
 
 void Adagrad::load(serialize::InputArchive& archive) {
-  serialize(*this, archive);
+  IValue pytorch_version;
+  if(archive.try_read("pytorch_version", pytorch_version)){
+    serialize(*this, archive);
+  }
+  else { // deserializing archives saved in old format (prior to version 1.5.0)
+    TORCH_WARN(
+      "Your serialized Adagrad optimizer is still using the old serialization format. ",
+      "You should re-save your Adagrad optimizer to use the new serialization format.");
+    std::vector<Tensor> sum_buffers;
+    std::vector<int64_t> step_buffers;
+    torch::optim::serialize(archive, "sum_buffers", sum_buffers);
+    torch::optim::serialize(archive, "step_buffers", step_buffers);
+    // since there were no param_groups prior to version 1.5.0, assuming all tensors are now in one param_group
+    std::vector<Tensor> params = param_groups_[0].params();
```

Comment:
It would be great to use `param_groups_.at(0)` instead of `param_groups_[0]`, because `.at` also does boundary checks.